### PR TITLE
feat(dashboard reporter): Retrieve the project name and version from the project under test using source link

### PR DESCRIPTION
**Pull request by 0xced**
Thanks to [SourceLink][1], the repository URL and the full version (including the git SHA1) of a project can be included in the produced assembly.

This commit uses the information computed by SourceLink to automatically retrieve the project name (github.com/organization/project) and project version which are required for the dashboard reporter.

With this technique, the table on the [Stryker dashboard documentation][2] can be updated with ✅ everywhere for the Stryker.NET column since it doesn't depend on a specific CI provider to provide the repository URL and version.

Note that I have run `Stryker.CLI --reporter progress --reporter html --reporter dashboard` on my local checkout of https://github.com/0xced/serilog-formatting-log4net/tree/20e86d55b4bd9e3682c3608530398e794df12481 and it automatically retrieved the project name (`github.com/0xced/serilog-formatting-log4net`) and project version (`1.0.0-rc.2+20e86d55b4bd9e3682c3608530398e794df12481`). The report was successfully uploaded to the [Stryker dashboard][3]

[1]: https://github.com/dotnet/sourcelink#readme
[2]: https://stryker-mutator.io/docs/General/dashboard/#retrieved-from-environment
[3]: https://dashboard.stryker-mutator.io/reports/github.com/0xced/serilog-formatting-log4net/1.0.0-rc.2%2B20e86d55b4bd9e3682c3608530398e794df12481

### CODE REVIEW

**Comment by 0xced** [comment]
I erroneously closed #1532 by deleting my branch and I was not able to reopen it (the reopen button would stay grayed out).

I have now resolved all conflicts and improved the way the project name and version are passed back to the dashboard reporter (no more hacky hidden StrykerOptions references).

@rouke-broersma It would be nice to merge this pull request before more conflicts occur. 😉

**Comment by rouke-broersma** [comment]
Sorry about that! I'll take a look when I get back from holiday tomorrow or Friday.

**Comment by 0xced** [comment]
No problem, enjoy your holiday! 🏖🌴

**Review by rouke-broersma — CHANGES_REQUESTED** [review]
I think all of the `GeneralStrykerException`s can be `InputException` since in all cases something is wrong with the input given to stryker. Other than that it looks good! :)

**Comment by 0xced** [comment]
Done in 7536c9e90b65be07f3b14ebd588866e2c8dae71b

File: src/Stryker.Core/Stryker.Core/Initialisation/InitialisationProcess.cs [hunk 1]
```diff
@@ -120,7 +120,7 @@ private void InitializeDashboardReporter(DashboardReporter dashboardReporter, Pr
 
                 if (!projectInfo.ProjectUnderTestAnalyzerResult.Properties.TryGetValue("TargetPath", out var targetPath))
                 {
-                    throw new StrykerInputException($"Can't read {subject.ToLowerInvariant()} because the TargetPath property was not found in {projectFilePath}");
```
Comment (0xced) [inline src/Stryker.Core/Stryker.Core/Initialisation/InitialisationProcess.cs]: Sorry, I messed my branches, cherry picked the commit and did not even verify if the code was still compiling. 😐

Comment (rouke-broersma) [reply src/Stryker.Core/Stryker.Core/Initialisation/InitialisationProcess.cs]: I fixed it for you :)

**Comment by AshleighAdams** [comment]
@0xced This change causes my project to break with:

```
Can't read project name and project version because the TargetPath property was not found in /home/runner/work/Verlite/Verlite/src/Verlite.Core/Verlite.Core.csproj
```

Is this not be a major breaking change? Or at least have upgrade instructions published somewhere?

**Comment by rouke-broersma** [comment]
Hi @AshleighAdams

This should AFAIK not have been a breaking change. If you used the dashboard reporter previously you would have set the required properties manually and those would still be used. If you turn on the dashboard reporter and don't manually set the required properties Stryker tries to collect them from sourcelink. If you had previously enabled the dashboard reporter not had not set up the required properties I think you would have gotten an error about missing information. Could you open an issue on the stryker-net repo with information about your use case? You're currently replying on a 6 month old closed issue for a docs change.

**Comment by 0xced** [comment]
@AshleighAdams Did you eventually figure it out? When I tried to run Stryker.NET version 1.3.0 with the dashboard reporter on a local checkout of Verlite I encountered no issues. Also, not finding the project version was the consequence of another problem from what I saw in the logs of your GitHub Actions run:
> [19:07:05 WRN] Buildalyzer could not find sourcefiles. This should not happen. Will fallback to filesystem scan. Please report an issue at github.

**Comment by AshleighAdams** [comment]
Hey @0xced, nope I didn't. I dropped the version just as a temporary workaround as I haven't had the time to look into it recently.

I also have the same as you though, where it runs with no issue on my local clone. The only differences I can think of are either a partial/shallow clone, the OS it's running on, or something regarding deterministic/reproducible builds being used when doing CI/CD builds (source paths are normalized, which could maybe affect things,). The last one sounds like the most likely candidate to me.

**Comment by AshleighAdams** [comment]
@0xced Yeah, my hunch was correct. Reproducible builds are what break the build. I'll create an issue and add some logs and such.

**Comment by AshleighAdams** [comment]
Issue is here: https://github.com/stryker-mutator/stryker-net/issues/1899

**Pull request merged**